contracts/FeeCustody.sol

```diff
@@ -0,0 +1,405 @@
+     * @param _oracle ASSET/USD ORACLE.
+     * @param _intermediaryPath path for univ3 swap.
+     * @param _poolFee fees for asset / distributionToken.
+
+     * If intermediary path then pool fee between both pairs
+     * (ex: AAVE / ETH , ETH / USDC)
+     * NOTE: if intermediaryPath empty then single hop swap
+     * OR remain unchanged if asset had already been added.
+     * NOTE: MUST BE ASSET / USD ORACLE
+     * NOTE: 3000 = 0.3% fee for pool fees
+     */
+    function setAsset(
+        address _asset,
+        address _oracle,
+        address[] calldata _intermediaryPath,
+        address[] calldata _poolFees
+    ) external onlyOwner {
+        require(_asset != address(0), "!address(0)");
+        uint8 _pathLen = _intermediaryPath.length;
+        uint8 _swapFeeLen = _poolFees.length;
+        uint8 _assetExists = assetExists[_asset];
+
+        // We must be setting new valid oracle, or want to keep as is if one exists
+        require(IChainlink(oracles[_asset]).decimals() == 8, "!ASSET/USD");
+        require(_pathLen < 2, "invalid intermediary path");
+        require(
+            _swapFeeLen > 0 && _swapFeeLen < 2,
+            "invalid pool fees array length"
+        );
+
+        // If not set asset
+        if (!_assetExists) {
+            assets[lastAssetIdx] = _asset;
+            ++lastAssetIdx;
+            assetExists[_asset] = true;
+        }
+
+        // Set oracle for asset
+        oracles[_asset] = _oracle;
+
+        // Multiple pool swaps are encoded through bytes called a `path`.
+        // A path is a sequence of token addresses and poolFees that define the pools used in the swaps.
+        // The format for pool encoding is (tokenIn, fee, tokenOut/tokenIn, fee, tokenOut)
+        // where tokenIn/tokenOut parameter is the shared token across the pools.
+        if (_pathLen > 0) {
+            intermediaryPath[_asset] = abi.encodePacked(
+                _asset,
+                _poolFees[0],
+                _intermediaryPath[0],
+                _poolFees[1],
+                address(distributionToken)
+            );
+        } else {
+            intermediaryPath[_asset] = abi.encodePacked(
+                _asset,
+                _poolFees[0],
+                address(distributionToken)
+            );
+        }
+
+        emit NewAsset(_asset, intermediaryPath[_asset]);
+    }
+
+    /**
+     * @notice
+     * recover all assets
+     * @dev Can be called by admin
+     */
+    function recoverAllAssets() external onlyOwner {
+        // For all added assets, if not removed, send to protocol revenue recipient
+        for (uint256 i = 0; i < lastAssetIdx; i++) {
+            _recoverAsset(assets[i]);
+        }
+    }
+
+    /**
+     * @notice
+     * recover specific asset
+     * @dev Can be called by admin
+     * @param _asset asset to recover
+     */
+    function recoverAsset(address _asset) external onlyOwner {
+        require(_asset != address(0), "!address(0)");
```

Comment:
Nit, this revert message should be `!asset` to distinguish it from the other address(0) checks